app/src/rgb_underglow.c

```diff
@@ -468,17 +468,31 @@ int zmk_rgb_underglow_change_spd(int direction) {
 
 #if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_AUTO_OFF_IDLE) ||                                          \
     IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_AUTO_OFF_USB)
-static int rgb_underglow_auto_state(bool *prev_state, bool new_state) {
-    if (state.on == new_state) {
+struct rgb_underglow_sleep_state {
+    bool is_awake;
+    bool rgb_state_before_sleeping;
+};
+
+static int rgb_underglow_auto_state(bool target_wake_state) {
+    static struct rgb_underglow_sleep_state sleep_state = {
+        is_awake : true,
+        rgb_state_before_sleeping : false
+    };
+
+    // wake up event while awake, or sleep event while sleeping -> no-op
+    if (target_wake_state == sleep_state.is_awake) {
         return 0;
     }
-    if (new_state) {
-        state.on = *prev_state;
-        *prev_state = false;
-        return zmk_rgb_underglow_on();
+    sleep_state.is_awake = target_wake_state;
+
+    if (sleep_state.is_awake) {
+        if (sleep_state.rgb_state_before_sleeping) {
+            return zmk_rgb_underglow_on();
+        } else {
+            return zmk_rgb_underglow_off();
+        }
     } else {
-        state.on = false;
-        *prev_state = true;
+        sleep_state.rgb_state_before_sleeping = sleep_state.on;
```

Comment:
Hey folks, I just ran a build of my firmware with underglow and its now failing. You can see the logs at https://github.com/jradtilbrook/zmk-config/actions/runs/8830708927/job/24244430141?check_suite_focus=true#step:11:143

Looks like the refactor is trying to reference a struct field that is not present to set the rgb status before sleeping. 

I think this might need to be updated to 
```suggestion
        zmk_rgb_underglow_get_state(&sleep_state.rgb_state_before_sleeping);
```

Reply:
Or maybe just
```suggestion
        sleep_state.rgb_state_before_sleeping = sleep.on;
```

Reply:
@petejohanson I've created #2284 to remedy this